src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/planning/SpaceBookingDayPlanner.kt

```diff
@@ -0,0 +1,147 @@
+    val bedLedger = BedLedger(beds)
+    val planned = mutableListOf<DayPlannerPlannedBooking>()
+    val unplanned = mutableListOf<DayPlannerUnplannedBooking>()
+
+    val sortedBookings = bookings.toList()
+      .sortedByDescending { b -> b.requiredCharacteristics.sumOf { c -> c.weighting } }
+      .toMutableList()
+
+    sortedBookings.forEach { booking ->
+      val requiresSingleRoom = booking.requiredCharacteristics.any { it.singleRoom }
+      val characteristicsExcludingSingleRoom = booking.requiredCharacteristics.filter { !it.singleRoom }.toSet()
+
+      when (
+        val findResult = bedLedger.findBed(
+          characteristics = characteristicsExcludingSingleRoom,
+          requiresSingleRoom = requiresSingleRoom,
+        )
+      ) {
+        is FindBedResult.BedsFound -> {
+          findResult.beds.forEach { bed ->
+            planned.add(
+              DayPlannerPlannedBooking(
+                bed = bed,
+                booking = booking,
+              ),
+            )
+            bedLedger.reserve(bed)
+          }
+        }
+
+        is FindBedResult.BedNotFound -> {
+          unplanned.add(
+            DayPlannerUnplannedBooking(
+              booking = booking,
+            ),
+          )
+        }
+      }
+    }
+
+    return DayPlannerResult(
+      planned = planned.toList(),
+      unplanned = unplanned.toSet(),
+    )
+  }
+}
+
+private class BedLedger(initialState: Set<Bed>) {
+  val allBeds = initialState.sortedBy { it.label }.toMutableList()
+  val availableBeds = initialState.sortedBy { it.label }.toMutableList()
+  val reservedBeds = mutableListOf<Bed>()
+
+  fun findBed(
+    characteristics: Set<Characteristic>,
+    requiresSingleRoom: Boolean,
+  ): FindBedResult {
+    val bedsWithMatchingCharacteristics =
+      if (requiresSingleRoom) {
+        availableBeds
+          .filter { availableBed -> availableBed.room.characteristics.containsAll(characteristics) }
+          .filter { availableBed -> !isRoomOccupied(availableBed) }
+          .sortedBy { availableBeds -> allBedsInRoom(availableBeds.room).size }
+      } else {
+        availableBeds
+          .filter { availableBed -> availableBed.room.characteristics.containsAll(characteristics) }
+      }
+
+    val bedWithLeastUnusedCharacteristics = bedsWithMatchingCharacteristics
+      .map { bed -> Pair(bed, bed.room.characteristics.minus(characteristics).size) }
+      .minByOrNull { it.second }?.first
+
+    return if (bedWithLeastUnusedCharacteristics != null) {
+      if (requiresSingleRoom) {
+        FindBedResult.BedsFound(allAvailableBedsInRoom(bedWithLeastUnusedCharacteristics.room))
+      } else {
+        FindBedResult.BedsFound(setOf(bedWithLeastUnusedCharacteristics))
+      }
+    } else {
+      FindBedResult.BedNotFound
+    }
+  }
+
+  fun reserve(bed: Bed) {
+    availableBeds.remove(bed)
+    reservedBeds.add(bed)
+  }
+
+  private fun allBedsInRoom(room: Room) = allBeds.filter { it.room == room }.toSet()
```

Comment:
Would it make sense to have this as a variable that is calculated once, then referred back to through each check, instead of a function?
e.g `private val bedsByRoom = allBeds.groupBy { it.room }`

Reply:
Whilst i tend to favour readability over optimisation i think this is a good optimisation that doesn't impact readability at all (actually probably makes it more readable). i've made the change 👍